document release checklist

esc opened the pull request:
This adds documentation and templates for the Numba release process.

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,97 @@
+    * [ ] post link to python-announce-list@python.org
+
+    ### Post Release:
+
+    * [ ] tag X.Y+1.0dev0 to start new development cycle on `master`
+    * [ ] update llvmlite dependency spec to match next version via PR to `master`
+    * [ ] update release checklist template
+    * [ ] close milestone (and then close this release issue)
+
+Second Releases Candidates, Final Releases and Patch Releases
+-------------------------------------------------------------
+
+A patch release usually involves a series of cherry-picks, so the recipe is
+slightly different.
+
+.. code-block::
+
+    ## numba X.Y.Z
+
+    * [ ] cherry-pick items from the X.Y.Z milestone into a PR
+    * [ ] merge change log modifications and cherry-picks to X.Y release branch
+    * [ ] https://github.com/numba/numba/pull/XXXX
+    * [ ] annotated tag X.Y.Z on release branch (no `v` prefix)
+    * [ ] build and upload conda packages on buildfarm (check "upload")
+    * [ ] build wheels (`$PYTHON_VERSIONS`) on the buildfarm
+    * [ ] upload wheels and sdist to PyPI (upload from `ci_artifacts`)
+    * [ ] verify packages uploaded to Anaconda Cloud and move to `numba/label/main`
+    * [ ] verify wheels for all platforms arrived on PyPi
+    * [ ] verify ReadTheDocs build
+    * [ ] clean up `ci_artifacts`
+    * [ ] send RC/FINAL announcement email / post announcement to discourse group
+    * [ ] post link to Twitter
+    * [ ] post link to python-announce-list@python.org
+
+    ### Post release
+
+    * [ ] cherry-pick change-log modifications to main branch (`master`)
+    * [ ] update release checklist template
+    * [ ] ping Anaconda Distro team to trigger a build for `defaults` (FINAL ONLY)
+    * [ ] close milestone (and then close this release issue)
```
Could these just be actual templates?

stuartarchibald replied:
In this directory: https://github.com/numba/numba/tree/master/.github/ISSUE_TEMPLATE

esc replied:
Maybe. However, I am not sure we can git version control the templates then. I think keeping track of how they evolve might be useful. Maybe we can update the templates from the documentation directly?

esc replied:
Maybe I can symlink the templates into Sphinx, let me try that.

stuartarchibald replied:
The templates are versioned by virtue of them being in the repo? You can see e.g. blame from numerous changes to the bug report template here:
https://github.com/numba/numba/blame/master/.github/ISSUE_TEMPLATE/Bug_report.md

or do you mean something else?

esc replied:
> The templates are versioned by virtue of them being in the repo? You can see e.g. blame from numerous changes to the bug report template here:
> https://github.com/numba/numba/blame/master/.github/ISSUE_TEMPLATE/Bug_report.md
> 
> or do you mean something else?

I wasn't aware that this is how tempaltes worked. I though they were simply attached to the repo in an unversioned form. Thank you for clearing this up.

stuartarchibald replied:
No problem. What's also quite nice about maybe doing it this way is that you don't have to expose the template in the https://github.com/numba/numba/blob/master/.github/ISSUE_TEMPLATE/config.yml (as it's not useful to users) but I think it's still accessible directly as needed. E.g. by accessing:
```
https://github.com/numba/numba/issues/new?template=THE_RELEASE_TEMPLATE.md
```

esc replied:
That is useful, I have included such URLs for both templates.

esc commented:
@stuartarchibald I have managed to convert the plain text templates to proper Github templates. I added the suffix (Maintainer Only) to the description for now, as I haven't yet found the documentation about hiding issue templates (I am not even sure it is possible).

esc commented:
@stuartarchibald I have trawled the Github documentation some more and have not discovered a setting for `config.yml` which allows us to hide specific issue templates and make them accessible only via the query parameters of a URL. I think the sufffix `(Maintainer Only)` may have to be "good enough for now". I could use `ALL CAPS` to emphasize this some more, but I'd rather we wait to resort to such extreme measures until it becomes absolutely necessary.

stuartarchibald commented on .github/ISSUE_TEMPLATE/sub_rc_checklist.yml:
```diff
@@ -0,0 +1,32 @@
+
+### Post release
+
+* [ ] cherry-pick change-log modifications to main branch (`master`)
```
This should go via a PR?

esc replied:
Yes, needs the addition "via PR".

esc replied:
done in 34f0949ee024a88f5efed01bc6f2c974d1942b5b

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+Numba Release Checklist Templates
+==================================
+
+The goal of the Numba release process -- from a birds eye perspective -- is to
```
```suggestion
The goal of the Numba release process -- from a high level perspective -- is to
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+==================================
+
+The goal of the Numba release process -- from a birds eye perspective -- is to
+publish source and binary artifacts that correspond to the given version
```
```suggestion
publish source and binary artifacts that correspond to a given version
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+between a Numba version and a corresponding llvmlite version.
+
+This section contains various notes and templates that can be used to create a
+Numba release checklist on the Numba Github issue tracker. This is the aid for
```
```suggestion
Numba release checklist on the Numba Github issue tracker. This is an aid for
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+the maintainers during the release process and helps to ensure that all tasks
+are completed in the correct order and that no tasks are accidentally omitted.
+
+Historically, the release checklists were copied from one release to the next.
+While this worked OKish for many years, often times tasks were lost during
+repeated copy and paste. As a result this document was created to keep track of
```
Is the history needed?

esc replied:
not sure, maybe not, it was mostly documenting my personal motivation, I can remove it if just adds overhead.

stuartarchibald replied:
I'm not sure it's necessary, a good reason for this process is presented in the previous paragraph. I'd rather spend effort keeping documentation such as this accurate with respect to changes/updates, if you feel differently am happy to discuss.

esc replied:
sure, no need for that perspective here.

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+While this worked OKish for many years, often times tasks were lost during
+repeated copy and paste. As a result this document was created to keep track of
+all potential tasks. All templates within this document are written in Github
+Markdown and are available as issue templates on Github. This means,
```
Why are they called `.yml` as files in the doc source tree if they are markdown `.md`?

esc replied:
Because the format of the templates is such that you write markdown for the template content and then wrap that in yaml to add metadata such as title etc...

esc replied:
So, this piece of documentation threw me off:

```
Issue templates are stored on the repository's default branch, in a hidden .github/ISSUE_TEMPLATE directory. If you create a template in another branch, it will not be available for collaborators to use. Issue template filenames are not case sensitive, and need a .md extension. To be included in the community profile checklist, issue templates must be located in the .github/ISSUE_TEMPLATE folder and contain valid name: and about: YAML front matter.
```

From: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates 

Will rename the files to `.md`.

esc replied:
done

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+Primary Release Candidate Checklist
+-----------------------------------
+
+This is for the first/primary release candidate for minor release. I.e. the
```
```suggestion
This is for the first/primary release candidate for minor release i.e. the
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+
+This is for the first/primary release candidate for minor release. I.e. the
+first release of every series. It is special, because during this release, the
+release branch will have to be created. Release candidate indexing begins at 0.
```
```suggestion
release branch will have to be created. Release candidate indexing begins at 1.
```
it begins at `1`? See note above about consistency.

esc replied:
Yes, it begins at 1.

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+    :lines: 9-
+
+
+Second Releases Candidates, Final Releases and Patch Releases
+-------------------------------------------------------------
```
```suggestion
Subsequent Release Candidates, Final Releases and Patch Releases
----------------------------------------------------------------
```

esc replied:
Yes, one of my favorite words!

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+Second Releases Candidates, Final Releases and Patch Releases
+-------------------------------------------------------------
+
+A patch release usually involves a series of cherry-picks, so the recipe is
+slightly different.
```
```suggestion
Releases subsequent to the first release in a series usually involves a series
of cherry-picks, the recipe is therefore slightly different.
```

stuartarchibald reviewed (commented):
Thanks for the patch, great to see this recorded and made easy to do.

esc commented on docs/source/developer/first_rc_checklist.yml:
```diff
@@ -0,0 +1 @@
+../../../.github/ISSUE_TEMPLATE/first_rc_checklist.yml
```
I am not sure what this comment refers to.

stuartarchibald replied:
Am wondering what this file is for, given it's just `literalinclude`d later on, could the file to which this file refers just be `literalinclude`d instead?

esc replied:
hmmmm, I think I am missing something here 🤷‍♂️

stuartarchibald replied:
I was wondering why not just include the file, but it looks like it later gets adjusted to create a better rendered output.

esc commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,51 @@
+first release of every series. It is special, because during this release, the
+release branch will have to be created. Release candidate indexing begins at 0.
+
+.. literalinclude:: first_rc_checklist.yml
+    :language: md
+    :lines: 9-
```
My intent here was to make the document readable and self-contained. Also the actual template contains additional "cruft" such as the yaml header which may be confusing, perhaps.

esc replied:
I thought about this some more, and I would like the templates to be rendered in the documentation. It makes it easier to refer to them and to view them. I have also added a link to open such an issue checklist, for convenience.

stuartarchibald replied:
I'm not entirely sure about this as the gains aren't yet clear to me, however I think it's worth a try to see how it goes.

esc commented on .github/ISSUE_TEMPLATE/first_rc_checklist.yml:
```diff
@@ -0,0 +1,36 @@
+---
+name: First Release Candidate Checklist (maintainer only)
+about: Checklist template for the first release of every series
+title: Numba X.Y.Zrc0 Checklist (FIXME)
```
You are correct. Will fix.

esc commented on .github/ISSUE_TEMPLATE/first_rc_checklist.yml:
```diff
@@ -0,0 +1,36 @@
+
+* [ ] tag X.Y+1.0dev0 to start new development cycle on `master`
+* [ ] update llvmlite dependency spec to match next version via PR to `master`
+* [ ] update release checklist template
```
The idea is to "port" any additional items that may have cropped up during a specific release  into the template so that they are available for the next release. Sort-of a continuous improvement of the template. Maybe it would be worth re-wording this bullet point text to make the intent more clear?

esc replied:
updated in b153c849833c023e3a938b1c75d96e02df31039

stuartarchibald replied:
thanks

esc commented on .github/ISSUE_TEMPLATE/sub_rc_checklist.yml:
```diff
@@ -0,0 +1,32 @@
+
+* [ ] cherry-pick items from the X.Y.Z milestone into a PR
+* [ ] merge change log modifications and cherry-picks to X.Y release branch
+* [ ] https://github.com/numba/numba/pull/XXXX
```
yes

esc replied:
done in 98a8bef4fbe609cac1be7edee1d28ebca8260cb9

esc commented:
@stuartarchibald thank you for your review. I have addressed most issues and left a few comments on items where we need to find consensus. Happy to see this moving closer to the finish line.

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,49 @@
+are completed in the correct order and that no tasks are accidentally omitted.
+
+If new or additional items do appear during release, please do remember to add
+them to the checklist templates. Note also, that the release process itself is
+always a work in progress. This means, some of the information here may be
+outdated. If you detect this, please do remember to submit a pull-request to
+update this document.
```
```suggestion
them to the checklist templates. Also note that the release process itself is
always a work in progress. This means that some of the information here may be
outdated. If you notice this please do remember to submit a pull-request to
update this document.
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,49 @@
+outdated. If you detect this, please do remember to submit a pull-request to
+update this document.
+
+All release checklists are available as Gitub issue templates, so to create a
+new release checklist, simply open a new issue and select the correct template.
```
```suggestion
All release checklists are available as Gitub issue templates. To create a new
release checklist simply open a new issue and select the correct template.
```

stuartarchibald commented on .github/ISSUE_TEMPLATE/first_rc_checklist.md:
```diff
@@ -0,0 +1,37 @@
+    - [ ] "remaining Pull-Requests from milestone"
+* [ ] merge change log changes
+    - [ ] "PR with changelog entries"
+* [ ] Create X.Y release branch
```
```suggestion
* [ ] create X.Y release branch
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,49 @@
+Numba Release Checklist Templates
+=================================
```
```suggestion
Numba Release Process
=====================
```

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,49 @@
+    :language: md
+    :lines: 9-
+
+Open a release checklist: https://github.com/numba/numba/issues/new?template=first_rc_checklist.md
```
```suggestion
`Open a primary release checklist <https://github.com/numba/numba/issues/new?template=first_rc_checklist.md>`_.
```
Can these just be standard links?

stuartarchibald commented on docs/source/developer/release.rst:
```diff
@@ -0,0 +1,49 @@
+    :language: md
+    :lines: 9-
+
+Open a release checklist: https://github.com/numba/numba/issues/new?template=sub_rc_checklist.md
```
```suggestion
`Open a subsequent release checklist <https://github.com/numba/numba/issues/new?template=sub_rc_checklist.md>`_.
```

stuartarchibald reviewed (commented):
Thanks for the updates. Few minor suggestions else looks good.

stuartarchibald reviewed (approved):
Thanks for all your efforts on this, patch looks good!

esc commented:
@stuartarchibald thank you for the feedback, I have included all suggestions now and also added the full-stops and capitals as per our OOB conversation.
